File: Puzzle15/MainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Puzzle15
{
    public partial class MainForm : Form
    {
        private Puzzle _puzzle;

        public MainForm()
        {
            InitializeComponent();
        }

        private void MainForm_Load(object sender, EventArgs e)
        {
            _puzzle = new Puzzle();
            _puzzle.Shuffle();
            _puzzle.ImageUpdated += (_, __) =>
            {
                PctPuzzle.Image = null;
                PctPuzzle.Image = _puzzle.Image;
            };
            PctPuzzle.Image = _puzzle.Image;
        }

        private void PctPuzzle_MouseDown(object sender, MouseEventArgs e)
        {
            var targetCell = new Point((int)(e.Location.X / (double)Puzzle.CellPixelWidth),
                          (int)(e.Location.Y / (double)Puzzle.CellPixelHeight));
            if (_puzzle.CanMove(targetCell))
            {
                _puzzle.MoveCell(targetCell);
            }
            if (_puzzle.IsCompleted())
            {
                PctPuzzle.Enabled = false;
                MessageBox.Show("おめでとうございます ！！！", "Puzzle 15", MessageBoxButtons.OK, MessageBoxIcon.Information);
            }
        }

        private void BtnShuffle_Click(object sender, EventArgs e)
        {
            _puzzle.Shuffle();
            PctPuzzle.Enabled = true;
        }
    }
}

File: Puzzle15VB/MainForm.vb
Public Class MainForm

    Private _puzzle As Puzzle

    Private Sub MainForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        _puzzle = New Puzzle()
        _puzzle.Shuffle()
        AddHandler _puzzle.ImageUpdated, AddressOf PuzzleImageUpdated
        PctPuzzle.Image = _puzzle.Img
    End Sub

    Private Sub BtnClear_Click(sender As Object, e As EventArgs) Handles BtnClear.Click
        _puzzle.Shuffle()
        PctPuzzle.Enabled = True
    End Sub

    Private Sub PuzzleImageUpdated(sender As Object, e As EventArgs)
        PctPuzzle.Image = Nothing
        PctPuzzle.Image = _puzzle.Img
    End Sub

    Private Sub PctPuzzle_MouseDown(sender As Object, e As MouseEventArgs) Handles PctPuzzle.MouseDown
        Dim targetCell = New Point(Math.Floor(e.Location.X / Puzzle.CellPixelWidth), Math.Floor(e.Location.Y / Puzzle.CellPixelHeight))
        Debug.WriteLine(targetCell)
        If _puzzle.CanMove(targetCell) Then
            _puzzle.MoveCell(targetCell)
        End If
        If _puzzle.IsCompleted() Then
            PctPuzzle.Enabled = False
            MessageBox.Show("おめでとうございます ！！！", "Puzzle 15", MessageBoxButtons.OK, MessageBoxIcon.Information)
        End If
    End Sub
End Class
